### PR TITLE
daml-lf: add GenMap to archive proto

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the state of #2256.

We add GenMap type and builtins to the archive proto.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -307,6 +307,16 @@ final case class Compiler(packages: PackageId PartialFunction Package) {
               case BMapToList => SBMapToList
               case BMapSize => SBMapSize
 
+              // GebMap
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]:
```suggestion
              // GenMap
```

File: daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto [hunk 2]
```diff
@@ -269,6 +269,10 @@ enum PrimType {
   // *Available in versions >= 1.3*
   MAP = 16;
 
+  // Builtin type 'TGenMap`
+  // *Available in versions >= 1.dev*
+  GENMAP = 20;
```
Comment (hurryabit) [inline daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto]: I suggest we move this down below `TYPE_REP` to keep the numbers in order.

Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto]: fixed

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 3]
```diff
@@ -125,6 +125,50 @@ private[validation] object Typing {
...
+          alpha.name -> KStar,
+          TForall(
+            beta.name -> KStar,
+            alpha ->: beta ->: TGenMap(alpha, beta) ->: TGenMap(alpha, beta))),
+      BGenMapLookup ->
+        TForall(
+          alpha.name -> KStar,
+          TForall(
+            beta.name -> KStar,
+            TText ->: TGenMap(alpha, beta) ->: TOptional(beta)
```
Comment (gerolf-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Should `TText` not rather be `alpha`?

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 4]
```diff
@@ -125,6 +125,50 @@ private[validation] object Typing {
...
+          )),
+      BGenMapDelete ->
+        TForall(
+          alpha.name -> KStar,
+          TForall(
+            beta.name -> KStar,
+            TText ->: TGenMap(alpha, beta) ->: TGenMap(alpha, beta)
```
Comment (gerolf-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Should `TText` not rather be `alpha`?

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Thanks ! fixed

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Thanks. fixed also.

**Pull request merged**